Remove stale values when updating solr index

erinc1915 opened the pull request:
Connected to issue https://github.com/JanusGraph/janusgraph/issues/1050
-This is a relatively minor change that simply adds a config option to not preserve stale values within SolrIndex's mutate function. 

Signed-off-by:  Erin Cornett erin.cornett@ibm.com

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](https://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [ ] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?
- [ ] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

mbrukman commented:
@erinc1915, thank you for your contribution to the JanusGraph project!

Looks like you're using a combination of usernames: @erinc1915 as well as @erinc-ibm (I'm looking at the [commits tab][commits-tab] which shows this). Note that @pluradj only added @erinc1915 to the CLA, so please either ensure that only that username is the one that's used for commits and PRs, or chat with @pluradj and ask him to update the CLA to switch the username, or add both usernames, depending on the situation.

Also, please ensure that your work email address is included in your GitHub profile (and validated); it does not have to be made public on your profile, it just has to be confirmed with GitHub that you own that email address.

If, after having made the above changes, @janusgraph-bot still says this PR is not passing the CLA check, please tag me into this PR again and I will look into it.

[commits-tab]: https://github.com/JanusGraph/janusgraph/pull/1547/commits

erinc1915 commented:
Hi @mbrukman , thank you for your help! I fixed the authoring issue (removed erinc-ibm) and made sure my work email address is verified on my GitHub account, but I am still not passing the CLA check.
Is there a step I'm missing?

porunov commented:
@erinc1915 could you please squash your commits into a single one?

erinc1915 commented:
> @erinc1915 could you please squash your commits into a single one?
Done!

erinc1915 commented:
Updated to include unit testing.

porunov commented on janusgraph-solr/src/test/java/org/janusgraph/diskstorage/solr/SolrIndexTest.java:
```diff
@@ -82,6 +89,9 @@ protected Configuration getLocalSolrTestConfig() {
 
         config.set(SolrIndex.ZOOKEEPER_URL, SolrRunner.getZookeeperUrls(), index);
         config.set(SolrIndex.WAIT_SEARCHER, true, index);
+        if(testingStaleValues) {
```
Didn't yet check the test but it looks confusing. Method `@BeforeAll` is executed only once before all tests. So, it always will be false here.
Instead, you should probably use parametrized test.

erinc1915 replied:
I'm confused by what you mean here- `testingStaleValues` is initialized as false outside of the `@BeforeAll ` method. It is set as true within the one function (`testUpdateListWithNoStaleValues`) that uses it, but is otherwise false because I didn't want to affect other tests with the altered config. 
Within `testUpdateListWithNoStaleValues`, SolrIndex is created with the remove-stale-values flag marked as true when `index = new SolrIndex(getLocalSolrTestConfig());` and `clopen()` are called.

porunov replied:
Sorry, I didn't check the code properly. Now I get what you want to do here.

erinc1915 commented:
I added a few comments to the code clarifying the tests I added in `SolrIndexTest.java`, and also explicitly scoped `boolean testingStaleValues` as `private` as per Codacity's code quality standards.

Thank you for the feedback and assistance so far!

porunov commented on janusgraph-solr/src/test/java/org/janusgraph/diskstorage/solr/SolrIndexTest.java:
```diff
@@ -175,4 +186,65 @@ public void testMapKey2Field_MappingSpaces() {
     }
+    
+    @Test
+    public void testUpdateListWithNoStaleValues() throws Exception {
```
The tests are mostly identical. I would better to write a single parametrized test for stale values check. Something like:
```
public static Stream<Boolean> collectionCardinalityReplaceStaleValue() {
    return ImmutableSet.of(true, false).stream();
}

@ParameterizedTest
@MethodSource("collectionCardinalityReplaceStaleValue")
public void testCollectionCardinalityStaleValues(Boolean replaceStaleValue) throws BackendException {
    final String store = "vertex";
    final String docid = "docid";
    final String defText = "1";
    final String revisedText = "2";

    if(replaceStaleValue) {
        testingStaleValues = true;
        index = new SolrIndex(getLocalSolrTestConfig());
    }

    initialize(store);
    Multimap<String, Object> initialProps = ImmutableMultimap.of(PHONE_LIST, defText);
    add(store, docid, initialProps, true);
    clopen();

    // Sanity check
    assertEquals(docid, tx.queryStream(new IndexQuery(store, PredicateCondition.of(PHONE_LIST, Cmp.EQUAL, "1"))).findFirst().get());

    tx.add(store, docid, PHONE_LIST, revisedText, false);
    tx.commit();
    clopen();

    if(replaceStaleValue) {
        // Should no longer return old text
        assertEquals(Optional.empty(), tx.queryStream(new IndexQuery(store, PredicateCondition.of(PHONE_LIST, Cmp.EQUAL, defText))).findFirst());
    } else {
        // Without the replace-stale-values flag, the old value should still be present
        assertEquals(docid, tx.queryStream(new IndexQuery(store, PredicateCondition.of(PHONE_LIST, Cmp.EQUAL, defText))).findFirst().get());
    }

    // the new value should also be present
    assertEquals(docid, tx.queryStream(new IndexQuery(store, PredicateCondition.of(PHONE_LIST, Cmp.EQUAL, revisedText))).findFirst().get());

    testingStaleValues = false;
}
```

erinc1915 replied:
Thank you for the suggestion, I updated the testing to match the example you showed above and it works as expected. I appreciate it!

chupman commented:
I'm not really sure how to address this, but currently we have multiple pull requests trying to fix index mutation consolidation for elasticsearch and solr. I went to put a second approval on this and found it to be quite similar to #1575, just with slightly different approaches. I spent a few hours going over JanusGraph index mutation consolidations today to make sure my hunch was correct. 

#1575 fixes the consolidation behavior but updating the way the collections are stored in the additions and deletions list to `field.value`, instead of just `field` if the Cardinality isn't SINGLE. As a result deletions don't get wiped out when there's an addition to the same field, unless they have the same value as well. 

This PR and #1588 add a config option that will tell solr/es to set rather than add, giving users the option of bypassing consolidation. I tend to think the behavior of using `set` instead of `add` makes more sense and it might be better to make it the default rather than a config option.

Both have CLAs, tests, and well written code. I'm pinging @FlorianHockmann, @porunov, @mbrukman, and @pluradj  to get some extra input on how we should proceed.

chupman commented:
I reached out today and was told that #1575 did not fix the issue this group was hitting. @erinc1915 
Please rebase your changes on 0.3 and update the target branch. 

I'll do a review on this at #1588 once they are retargeted.

porunov commented:
I am just reviewing the code again and started to think if this code is OK or not.
I haven't tested my thought and I am not sure if I will have time till Monday (but I will try).

There are some questions right now:
- If we enable `replace-stale-values` how can we add a new value to the LIST or SET without removing all other values from the collection?
- When we enable `replace-stale-values` and assume we have an initial LIST collection with an element "Hello". Then we add a new element in another transaction with an element "World". With the enabled property `replace-stale-values` we will rewrite the collection "Hello" with the collection "World". So, we will have only "World" in Solr. But will it remove "Hello" from the vertex in Cassandra? If not, then this solution isn't right because our index will have only "World" and out vertex will have "Hello", "World". 

I didn't check it. It is just questions right now. I will try to test these thoughts a little bit later.

erinc1915 commented:
@porunov You raise valid points, this configuration option is only useful in certain implementations. I tried to add clarification to my documentation in solr.adoc and elasticsearch.adoc to explain this as well. The `remove-stale-indexes` flag would be used in implementations where the entire attribute list of a multivalued property is updated in one transaction.

For the 'Hello World' example, it would be assumed that the original attribute list is `{"Hello"}` and the new input is the list `{"Hello", "World"}`. 
If you are incrementally adding single attributes to the list, you would not want to use this flag.

It might make more sense in the future to create a list-overwriting `update` mutation separate from `addition`, but for the upcoming release I think this is a viable solution.

porunov commented:
@erinc1915 Am I right that enabling this property to `true` may lead to inconsistency between storage server and search server (for example with "Hello" "World" example where we have only "World" in Solr but "Hello" "World" in Cassandra)?
If so, this option is usable only for a very limited number of cases and it is very dangerous to use. 
In case when we have a list `{"Hello"}` in out database and would like to make a new list like this `{"World"}` then we should perform 2 separate transactions:
1. Remove `"Hello"` from the databse (so that it would be removed from Solr and Cassandra).
2. Add `"World"` into the database.

Notice, if we don't remove `"Hello"` in a first transaction then we will get a permanent inconsistency between our Solr and Cassandra. I.e. Solr will have `{"World"}` and Cassandra will have `{"Hello", "World"}`. 

Am I right about the above scenario or not? I didn't yet check it

porunov commented:
This solution doesn't work. The explanation is here: https://github.com/JanusGraph/janusgraph/pull/1588#issuecomment-497293677

Also, notice that we have merged #1606 which fixes problems with deletions in the same traverse.
